From 4968e8fbe8283834351d78be09ebcf4cc8416bf1 Mon Sep 17 00:00:00 2001
From: Hari Chandrakanthan <quic_haric@quicinc.com>
Date: Tue, 20 Feb 2024 17:12:21 +0530
Subject: [PATCH] wifi: ath12k: add support to find peer using ast

Add support to find a peer in the list of available peers
using ast hash.

For MU case, ppdu_info structure does not contain peer id for
each user. The current logic tries to find peer using same peer_id
for all the users. This logic is incorrect.

Fix it by using ast_index for peer search.

Signed-off-by: Hari Chandrakanthan <quic_haric@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_mon.c |  3 +--
 drivers/net/wireless/ath/ath12k/peer.c   | 14 ++++++++++++++
 drivers/net/wireless/ath/ath12k/peer.h   |  1 +
 3 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp_mon.c b/drivers/net/wireless/ath/ath12k/dp_mon.c
index 3a05c1e..3c7f770 100644
--- a/drivers/net/wireless/ath/ath12k/dp_mon.c
+++ b/drivers/net/wireless/ath/ath12k/dp_mon.c
@@ -3569,8 +3569,7 @@ ath12k_dp_mon_rx_update_user_stats(struct ath12k *ar,
 	if (ppdu_info->peer_id == HAL_INVALID_PEERID)
 		return;
 
-	peer = ath12k_peer_find_by_id(ar->ab, ppdu_info->peer_id);
-
+	peer = ath12k_peer_find_by_ast(ar->ab, user_stats->ast_index);
 	if (!peer) {
 		ath12k_warn(ar->ab, "peer with peer id %d can't be found\n",
 			    ppdu_info->peer_id);
diff --git a/drivers/net/wireless/ath/ath12k/peer.c b/drivers/net/wireless/ath/ath12k/peer.c
index 89f5ca2..0075ff3 100644
--- a/drivers/net/wireless/ath/ath12k/peer.c
+++ b/drivers/net/wireless/ath/ath12k/peer.c
@@ -132,6 +132,20 @@ struct ath12k_peer *ath12k_peer_find_by_id(struct ath12k_base *ab,
 	}
 }
 
+struct ath12k_peer *ath12k_peer_find_by_ast(struct ath12k_base *ab,
+					    int ast_hash)
+{
+	struct ath12k_peer *peer;
+
+	lockdep_assert_held(&ab->base_lock);
+
+	list_for_each_entry(peer, &ab->peers, list)
+		if (ast_hash == peer->ast_hash)
+			return peer;
+
+	return NULL;
+}
+
 bool ath12k_peer_exist_by_vdev_id(struct ath12k_base *ab, int vdev_id)
 {
 	struct ath12k_peer *peer;
diff --git a/drivers/net/wireless/ath/ath12k/peer.h b/drivers/net/wireless/ath/ath12k/peer.h
index 4e06d97..105f174 100644
--- a/drivers/net/wireless/ath/ath12k/peer.h
+++ b/drivers/net/wireless/ath/ath12k/peer.h
@@ -105,6 +105,7 @@ struct ath12k_peer *ath12k_peer_find_list_by_id(struct ath12k_base *ab,
 struct ath12k_peer *ath12k_peer_find_by_addr(struct ath12k_base *ab,
 					     const u8 *addr);
 struct ath12k_peer *ath12k_peer_find_by_id(struct ath12k_base *ab, int peer_id);
+struct ath12k_peer *ath12k_peer_find_by_ast(struct ath12k_base *ab, int ast_hash);
 void ath12k_peer_cleanup(struct ath12k *ar, u32 vdev_id);
 int ath12k_peer_delete(struct ath12k *ar, u32 vdev_id, u8 *addr);
 int ath12k_ml_link_peers_delete(struct ath12k_vif *ahvif, struct ath12k_sta *ahsta);
-- 
2.34.1

